### PR TITLE
ISPN-8227 OptimisticTxPartitionAndMergeDuringPrepareTest.testPrimaryOwnerIsolatedPartition fails randomly

### PR DESCRIPTION
**Pull request by rvansa**
https://issues.jboss.org/browse/ISPN-8227

### CODE REVIEW

**Comment by tristantarrant** [comment]
Its counterpart (duringCommit) failed at least once in CI: http://ci.infinispan.org/job/Infinispan/job/PR-5408/10/testReport/junit/org.infinispan.partitionhandling/OptimisticTxPartitionAndMergeDuringCommitTest/testOriginatorIsolatedPartitionWithDiscard_DIST_SYNC_/

Can the fix be applied to both ?

**Comment by rvansa** [comment]
@tristantarrant Should be, IMO... I'll try to reproduce it locally.

**Comment by rvansa** [comment]
Hmm, the original test failed every now and then on my machine, but I can't get the failure from CI.

**Comment by pruivo** [comment]
aren't you hitting https://issues.jboss.org/browse/ISPN-8232? it is going to require some work...

File: core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java [hunk 1]
```diff
@@ -471,10 +473,17 @@ public void remoteTransactionCommitted(GlobalTransaction gtx, boolean onePc) {
    }
 
    public final RemoteTransaction removeRemoteTransaction(GlobalTransaction txId) {
-      RemoteTransaction removed = remoteTransactions.remove(txId);
-      if (trace) log.tracef("Removed remote transaction %s ? %s", txId, removed);
-      releaseResources(removed);
-      return removed;
+      ByRef<RemoteTransaction> removed = new ByRef<>(null);
+      // we need to mark the transaction inside compute() even if it does not exist
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java]: Why not mark the tx as completed before removing it from the map?

**Comment by rvansa** [comment]
@pruivo That might be the cause for the other failure; here I've identified a race in data structures access and I am fixing that one.

File: core/src/main/java/org/infinispan/util/concurrent/BlockingTaskAwareExecutorServiceImpl.java [hunk 2]
```diff
@@ -167,7 +168,15 @@ public void run() {
             ArrayDeque<BlockingRunnable> readyList = new ArrayDeque<>(size);
             for (Iterator<BlockingRunnable> iterator = blockedTasks.iterator(); iterator.hasNext(); ) {
                BlockingRunnable runnable = iterator.next();
-               if (runnable.isReady()) {
+               boolean ready;
+               try {
+                  ready = runnable.isReady();
+               } catch (Exception e) {
+                  log.debugf(e, "Failed to check ready state of %s, dropping.", runnable);
```
[the remark this answers is not in the document]
Comment (rvansa) [reply core/src/main/java/org/infinispan/util/concurrent/BlockingTaskAwareExecutorServiceImpl.java]: Would you like me to add a `getCommand()` method to `BlockingRunnable` iface? I've rather added `toString()` impl to `DefaultTopologyRunnable`; I could add it to other implementing classes as well.

File: core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java (hunk 1 above)
Comment (rvansa) [reply core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java]: It seems I could; I thought I have an counterexample but probably I was mistaken.

Comment (rvansa) [reply core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java]: Actually I need to see if the tx was `successful`; trying to mark before remove/compute means that I'd need to do another lookup into `remoteTransaction` anyway.

File: core/src/main/java/org/infinispan/util/concurrent/BlockingTaskAwareExecutorServiceImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (danberindei) [reply core/src/main/java/org/infinispan/util/concurrent/BlockingTaskAwareExecutorServiceImpl.java]: Adding `toString()` to all the `BlockingRunnable` impls is fine.

File: core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java (hunk 1 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java]: I missed that, let's keep it in `compute()` then.

**Comment by rvansa** [comment]
@danberindei Added those `toString()`s.

File: core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java [hunk 3]
```diff
@@ -370,7 +372,7 @@ private RemoteTransaction getOrCreateRemoteTransaction(GlobalTransaction globalT
 
       int viewId = rpcManager.getTransport().getViewId();
       if (!rpcManager.getTransport().getMembers().contains(globalTx.getAddress())) {
-         throw new IllegalStateException(
+         throw new CacheException(
```
Comment (tristantarrant) [inline core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java]: Possibly this deserves a localized message

File: core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java [hunk 4]
```diff
@@ -383,7 +385,7 @@ private RemoteTransaction getOrCreateRemoteTransaction(GlobalTransaction globalT
             return existing;
          } else {
             if (isTransactionCompleted(gtx)) {
-               throw new IllegalStateException("Cannot create remote transaction " + gtx + ", already completed");
+               throw new CacheException("Cannot create remote transaction " + gtx + ", already completed");
```
Comment (tristantarrant) [inline core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java]: Localized message ?

**Comment by rvansa** [comment]
Localized the exceptions.

**Comment by tristantarrant** [comment]
Merged, thanks

**Pull request merged**